File: mobile/apps/photos/plugins/ente_photos_platform/android/src/main/kotlin/io/ente/photos/platform/flutter/ThumbnailChannelAdapter.kt
package io.ente.photos.platform.flutter

import android.os.Handler
import android.os.Looper
import io.ente.photos.platform.media.MediaAssetKind
import io.ente.photos.platform.media.MediaLibraryError
import io.ente.photos.platform.media.MediaThumbnailFit
import io.ente.photos.platform.media.MediaThumbnailOutcome
import io.ente.photos.platform.media.MediaThumbnailRequest
import io.ente.photos.platform.media.MediaThumbnailService
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

internal class ThumbnailChannelAdapter : MethodChannel.MethodCallHandler {
    private val mainHandler = Handler(Looper.getMainLooper())
    private lateinit var channel: MethodChannel
    private lateinit var service: MediaThumbnailService
    private var attached = false

    fun attach(binding: FlutterPlugin.FlutterPluginBinding) {
        service = MediaThumbnailService(binding.applicationContext)
        channel = MethodChannel(binding.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)
        attached = true
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "thumbnail.load" -> load(call.arguments, result)
            "thumbnail.cancel" -> cancel(call.arguments, result)
            else -> result.notImplemented()
        }
    }

    fun detach() {
        attached = false
        channel.setMethodCallHandler(null)
        service.close()
        mainHandler.removeCallbacksAndMessages(null)
    }

    private fun load(arguments: Any?, result: MethodChannel.Result) {
        val values = strictArguments(arguments, LOAD_KEYS)
            ?: return result.error(MediaLibraryError.INVALID_REQUEST.channelValue, null, null)
        val operationId = values.string("operationId", 128)
        val request = runCatching {
            MediaThumbnailRequest(
                assetId = requireNotNull(values.string("assetId", 1024)),
                kind = when (values["kind"]) {
                    "image" -> MediaAssetKind.IMAGE
                    "video" -> MediaAssetKind.VIDEO
                    else -> error("Invalid media kind")
                },
                widthPx = requireNotNull(values.int("widthPx")),
                heightPx = requireNotNull(values.int("heightPx")),
                fit = when (values["fit"]) {
                    "contain" -> MediaThumbnailFit.CONTAIN
                    "cover" -> MediaThumbnailFit.COVER
                    else -> error("Invalid thumbnail fit")
                },
                quality = requireNotNull(values.int("quality")),
                allowNetworkAccess = requireNotNull(values["allowNetworkAccess"] as? Boolean),
            )
        }.getOrNull()
        if (operationId == null || request == null) {
            result.error(MediaLibraryError.INVALID_REQUEST.channelValue, null, null)
            return
        }
        service.load(operationId, request) { outcome ->
            mainHandler.post {
                if (!attached) return@post
                when (outcome) {
                    is MediaThumbnailOutcome.Success -> result.success(
                        mapOf(
                            "jpegBytes" to outcome.thumbnail.jpegBytes,
                            "widthPx" to outcome.thumbnail.widthPx,
                            "heightPx" to outcome.thumbnail.heightPx,
                        ),
                    )

                    is MediaThumbnailOutcome.Failure ->
                        result.error(outcome.error.channelValue, null, null)
                }
            }
        }
    }

    private fun cancel(arguments: Any?, result: MethodChannel.Result) {
        val values = strictArguments(arguments, CANCEL_KEYS)
        val operationId = values?.string("operationId", 128)
        if (operationId == null) {
            result.error(MediaLibraryError.INVALID_REQUEST.channelValue, null, null)
            return
        }
        service.cancel(operationId) {
            mainHandler.post {
                if (attached) result.success(null)
            }
        }
    }

    private fun strictArguments(arguments: Any?, keys: Set<String>): Map<*, *>? {
        val values = arguments as? Map<*, *> ?: return null
        if (values.keys.any { it !is String } || values.keys != keys) return null
        return values
    }

    private fun Map<*, *>.string(key: String, maximumBytes: Int): String? =
        (this[key] as? String)?.takeIf {
            it.isNotEmpty() && it.toByteArray(Charsets.UTF_8).size <= maximumBytes
        }

    private fun Map<*, *>.int(key: String): Int? = this[key] as? Int

    private val MediaLibraryError.channelValue: String
        get() = when (this) {
            MediaLibraryError.CANCELLED -> "cancelled"
            MediaLibraryError.PERMISSION_DENIED -> "permissionDenied"
            MediaLibraryError.ASSET_NOT_FOUND -> "assetNotFound"
            MediaLibraryError.RESOURCE_UNAVAILABLE -> "resourceUnavailable"
            MediaLibraryError.NETWORK_UNAVAILABLE -> "networkUnavailable"
            MediaLibraryError.UNSUPPORTED_FORMAT -> "unsupportedFormat"
            MediaLibraryError.INVALID_REQUEST -> "invalidRequest"
            MediaLibraryError.BUSY -> "busy"
            MediaLibraryError.UNSUPPORTED_PLATFORM -> "unsupportedPlatform"
            MediaLibraryError.PLATFORM_FAILURE -> "platformFailure"
        }

    private companion object {
        const val CHANNEL_NAME = "io.ente.photos.platform/media_library/commands.v1"
        val LOAD_KEYS = setOf(
            "operationId",
            "assetId",
            "kind",
            "widthPx",
            "heightPx",
            "fit",
            "quality",
            "allowNetworkAccess",
        )
        val CANCEL_KEYS = setOf("operationId")
    }
}

File: mobile/apps/photos/plugins/ente_photos_platform/ios/Flutter/ThumbnailChannelAdapter.swift
@preconcurrency import Flutter
import Foundation

@MainActor
final class ThumbnailChannelAdapter {
    private let channel: FlutterMethodChannel
    private let service = MediaThumbnailService()
    private var isAttached = true

    init(registrar: FlutterPluginRegistrar) {
        channel = FlutterMethodChannel(
            name: Self.channelName,
            binaryMessenger: registrar.messenger()
        )
        channel.setMethodCallHandler { [weak self] call, result in
            Task { @MainActor [weak self] in
                self?.handle(call, result: result)
            }
        }
    }

    func detach() {
        isAttached = false
        channel.setMethodCallHandler(nil)
        service.detach()
    }

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard isAttached else { return }
        switch call.method {
        case "thumbnail.load":
            load(arguments: call.arguments, result: result)
        case "thumbnail.cancel":
            cancel(arguments: call.arguments, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func load(arguments: Any?, result: @escaping FlutterResult) {
        guard let values = strictArguments(arguments, keys: Self.loadKeys),
            let operationID = string(values["operationId"], maximumBytes: 128)
        else {
            result(flutterError(.invalidRequest))
            return
        }
        let request: MediaThumbnailRequest
        do {
            guard let assetID = string(values["assetId"], maximumBytes: 1024),
                let kind = mediaKind(values["kind"]),
                let widthPx = int(values["widthPx"]),
                let heightPx = int(values["heightPx"]),
                let fit = thumbnailFit(values["fit"]),
                let quality = int(values["quality"]),
                let allowNetworkAccess = bool(values["allowNetworkAccess"])
            else {
                throw MediaLibraryError.invalidRequest
            }
            request = try MediaThumbnailRequest(
                assetID: assetID,
                kind: kind,
                widthPx: widthPx,
                heightPx: heightPx,
                fit: fit,
                quality: quality,
                allowNetworkAccess: allowNetworkAccess
            )
        } catch {
            result(flutterError(.invalidRequest))
            return
        }
        service.load(operationID: operationID, request: request) { [weak self] outcome in
            guard let self, isAttached else { return }
            switch outcome {
            case .success(let thumbnail):
                result([
                    "jpegBytes": FlutterStandardTypedData(bytes: thumbnail.jpegData),
                    "widthPx": thumbnail.widthPx,
                    "heightPx": thumbnail.heightPx,
                ])
            case .failure(let error):
                result(flutterError(error))
            }
        }
    }

    private func cancel(arguments: Any?, result: @escaping FlutterResult) {
        guard let values = strictArguments(arguments, keys: Self.cancelKeys),
            let operationID = string(values["operationId"], maximumBytes: 128)
        else {
            result(flutterError(.invalidRequest))
            return
        }
        service.cancel(operationID: operationID) { [weak self] in
            guard let self, isAttached else { return }
            result(nil)
        }
    }

    private func strictArguments(_ arguments: Any?, keys: Set<String>) -> [String: Any]? {
        guard let values = arguments as? [String: Any], Set(values.keys) == keys else { return nil }
        return values
    }

    private func string(_ value: Any?, maximumBytes: Int) -> String? {
        guard let value = value as? String, !value.isEmpty,
            value.lengthOfBytes(using: .utf8) <= maximumBytes
        else {
            return nil
        }
        return value
    }

    private func int(_ value: Any?) -> Int? {
        guard let number = value as? NSNumber,
            CFGetTypeID(number) != CFBooleanGetTypeID(),
            number.doubleValue.rounded(.towardZero) == number.doubleValue,
            number.int64Value >= Int64(Int.min),
            number.int64Value <= Int64(Int.max)
        else {
            return nil
        }
        return number.intValue
    }

    private func bool(_ value: Any?) -> Bool? {
        guard let number = value as? NSNumber, CFGetTypeID(number) == CFBooleanGetTypeID() else {
            return nil
        }
        return number.boolValue
    }

    private func mediaKind(_ value: Any?) -> MediaAssetKind? {
        switch value as? String {
        case "image": .image
        case "video": .video
        default: nil
        }
    }

    private func thumbnailFit(_ value: Any?) -> MediaThumbnailFit? {
        switch value as? String {
        case "contain": .contain
        case "cover": .cover
        default: nil
        }
    }

    private func flutterError(_ error: MediaLibraryError) -> FlutterError {
        FlutterError(code: error.channelValue, message: nil, details: nil)
    }

    private static let channelName = "io.ente.photos.platform/media_library/commands.v1"
    private static let loadKeys: Set<String> = [
        "operationId",
        "assetId",
        "kind",
        "widthPx",
        "heightPx",
        "fit",
        "quality",
        "allowNetworkAccess",
    ]
    private static let cancelKeys: Set<String> = ["operationId"]
}

private extension MediaLibraryError {
    var channelValue: String {
        switch self {
        case .cancelled: "cancelled"
        case .permissionDenied: "permissionDenied"
        case .assetNotFound: "assetNotFound"
        case .resourceUnavailable: "resourceUnavailable"
        case .networkUnavailable: "networkUnavailable"
        case .unsupportedFormat: "unsupportedFormat"
        case .invalidRequest: "invalidRequest"
        case .busy: "busy"
        case .unsupportedPlatform: "unsupportedPlatform"
        case .platformFailure: "platformFailure"
        }
    }
}
